src/main/java/org/opensearch/plugin/insights/core/service/QueryGrouper.java

```diff
@@ -0,0 +1,292 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.plugin.insights.core.service;
+
+import static org.opensearch.plugin.insights.settings.QueryInsightsSettings.TOP_N_QUERIES_MAX_GROUPS;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.PriorityQueue;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.common.collect.Tuple;
+import org.opensearch.plugin.insights.rules.model.AggregationType;
+import org.opensearch.plugin.insights.rules.model.Attribute;
+import org.opensearch.plugin.insights.rules.model.GroupingType;
+import org.opensearch.plugin.insights.rules.model.MetricType;
+import org.opensearch.plugin.insights.rules.model.SearchQueryRecord;
+import org.opensearch.plugin.insights.settings.QueryInsightsSettings;
+
+/**
+ * Handles grouping of search queries based on the GroupingType for the MetricType
+ * Following algorithm :
+ */
+public class QueryGrouper {
```

Comment:
We have one `TopQueriesService` per metricType : latency, cpu and memory. We have all the logic for the exporting query records, rotating window and so on in the TopQueriesService. The grouper is a layer on top of this service which leverages all the above. All it does is group queries by `group_id`. Everything else stays the same as TopQueriesService.

Reply:
I don't necessarily agree with the current design. My proposal was from a point of view that any store(min/max heap) should reside in a single place so that responsibility of classes are well defined. Currently minHeap store is being shared between TopQueriesService and QueryGrouper which seems wrong as ideally all those things should happen at a single place to ensure thread safety and logical code IMO.

As I see the code, you can easily segregate the logic and move minHeap store to a single place. Plus having different implementation of store also gives a user ability to easily write another implementation and plug that in.

OR other option is to have QueryGrouper only do some grouping logic(as its name suggests) and not perform any get/put operations on the queues. And keep rest of the stuff in service itself.

Reply:
As discussed added the following : https://github.com/opensearch-project/query-insights/pull/66#issuecomment-2322535233